### PR TITLE
Everyone's Got Problems (and error codes)

### PR DESCRIPTION
**Pull request by nmonterroso**
@Wikia/services-team 
https://wikia-inc.atlassian.net/browse/SERVICES-318

Changes the English error-messages to be descriptive error code strings. It'll be up to clients to turn those codes into user-facing translated strings. Also introducing the `Problem` struct as defined in the story, and the `Problematic` interface to convert other types to `Problem`, which `ErrorList` now implements. 

Note that we cannot change the `Content-Type` header right now because the osin library forcefully adds `Content-Type: application/json` to responses, and we'd end up with multiple `Content-Type` headers. See https://github.com/RangelReale/osin/pull/60

### CODE REVIEW

File: common/error_list.go [hunk 1]
```diff
@@ -1,9 +1,17 @@
...
+	"github.com/Wikia/helios/problem"
+)
 
 type ErrorList struct {
 	errors map[string][]error
 }
 
+const (
+	AdditionalField = "field"
+)
```
Comment (drsnyder) [inline common/error_list.go]: Does `error_list` need to know about this? I would expect something like `aProblem.AddAdditional(field)` instead of `aProblem.AddAdditional(AdditionalField, field)`.

File: common/error_list.go [hunk 2]
```diff
@@ -54,3 +62,23 @@ func (list *ErrorList) AsStrings() map[string][]string {
 
 	return errorMessages
 }
+
+func (list *ErrorList) AsProblem(title string) *problem.Problem {
+	errors := make([]*problem.Error, 0)
+
+	for field, errorList := range list.errors {
+		for _, error := range errorList {
+			errors = append(errors, problem.NewError(error.Error()).Additional(AdditionalField, field))
```
Comment (drsnyder) [inline common/error_list.go]: Is the `error.Error()` required? That is, can this be `problem.NewError().Additional(field)`?

File: controllers/helpers.go [hunk 3]
```diff
@@ -13,3 +15,10 @@ func originHandler(fn func(http.ResponseWriter, *http.Request)) http.HandlerFunc
 		fn(w, r)
 	}
 }
+
+func setOutputProblem(resp *osin.Response, p problem.Problematic, problemTitle string, statusCode int) {
+	// commented out since the osin library always adds content-type application/json https://github.com/RangelReale/osin/pull/60
+//	resp.Headers.Set("Content-Type", problem.ContentType)
+	resp.Output = p.AsProblem(problemTitle).AsMap()
+	resp.StatusCode = statusCode
+}
```
Comment (drsnyder) [inline controllers/helpers.go]: I think we should decouple our output from `osin`. It's convenient that it already provides an output generator but it's also coupling us to a 3rd party library that a) we might swap out later, b) we might remove entirely c) whose interface may change.

File: common/error_list.go (hunk 1 above)
Comment (nmonterroso) [inline common/error_list.go]: It does need to know about it. An additional is a map of key/vals that are specific to the type of error that's being reported. For these the map key is "field" and the value is the field of registration that failed, like "username" or "email". A different problem type could be something like:

```json
{
  "code": "something_went_wrong",
  "additional": {
    "count": 5,
    "other_key": "something else"
  }
}
```

Though maybe doesn't belong in `ErrorList` but in `User`

File: common/error_list.go (hunk 2 above)
Comment (nmonterroso) [inline common/error_list.go]: It could be, though it's never accessed as an `error` once it gets into the `problem.Error` struct, so I don't see value in doing that right now.

File: controllers/helpers.go (hunk 3 above)
Comment (nmonterroso) [inline controllers/helpers.go]: That sounds right, but that should probably be part of its own PR. Can you open a ticket to do that?

Comment (drsnyder) [inline controllers/helpers.go]: Yeah, I don't think it should be done as part of this PR. It feels a little like we are cobbling together a custom web framework in helios.

**Comment by nmonterroso** [comment]
@drsnyder I added `errorlist.UserValidation` which subtypes `ErrorList` and overwrites the `AsProblem` function, so now `ErrorList` doesn't know any details about the `additional` required when validation a user. Can you take a look?

File: controllers/register_controller_e2e_test.go [hunk 4]
```diff
@@ -165,3 +155,18 @@ func TestE2eRegisterTestSuite(t *testing.T) {
...
+		if additional[errorlist.ValidationAdditionalField] == field {
+			return true
+		}
+	}
+
+	return false
+}
+
+func pseudoRandomString(base string, purpose string) string {
+	return base + purpose + strconv.FormatInt(time.Now().Unix(), 10)
```
Comment (drsnyder) [inline controllers/register_controller_e2e_test.go]: Minor: Want to use `rand.Int63()` from `math/rand`?

File: problem/problem.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+	for _, error := range problem.errors {
+		errors = append(errors, error.AsMap())
+	}
+
+	return map[string]interface{} {
+		Description: problem.description,
+		Errors: errors,
+	}
+}
```
Comment (drsnyder) [inline problem/problem.go]: We probably mean `title` in here rather than `description`. The `description` field is not in the [draft](https://tools.ietf.org/html/draft-nottingham-http-problem-07#section-3.1).

File: errorlist/error_list.go [hunk 6]
```diff
@@ -54,3 +59,23 @@ func (list *ErrorList) AsStrings() map[string][]string {
...
+	errors := make([]*problem.Error, 0)
+
+	for _, errorList := range list.errors {
+		for _, error := range errorList {
+			errors = append(errors, problem.NewError(error.Error()))
+		}
+
+	}
+
+	return problem.NewProblem("An unexpected error occurred", errors)
```
Comment (drsnyder) [inline errorlist/error_list.go]: :+1: we are one step closer to symmetry with the [Java interface in pandora](https://github.com/Wikia/pandora/blob/b12fa2a62869e795b9a35f385d553b62531eece0/lib/service-core/src/main/java/com/wikia/pandora/http/StandardResponses.java.

Comment (drsnyder) [inline errorlist/error_list.go]: Not that symmetry should necessary be a goal, but I think it will lower the cognitive burden when switching back and forth.

File: controllers/register_controller_e2e_test.go (hunk 4 above)
Comment (nmonterroso) [inline controllers/register_controller_e2e_test.go]: 03f0a4b

File: controllers/helpers.go [hunk 7]
```diff
@@ -13,3 +15,10 @@ func originHandler(fn func(http.ResponseWriter, *http.Request)) http.HandlerFunc
 		fn(w, r)
 	}
 }
+
+func setOutputProblem(resp *osin.Response, p problem.Problematic, problemTitle string, statusCode int) {
```
Comment (drsnyder) [inline controllers/helpers.go]: :boom: now can have helpers like `NotFound()`, `Conflict()`, etc.

File: problem/problem.go (hunk 5 above)
Comment (nmonterroso) [inline problem/problem.go]: 8746829

File: problem/problem.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+func (problem *Problem) AsMap() map[string]interface{} {
+	errors := make([]map[string]interface{}, 0)
+
+	for _, error := range problem.errors {
+		errors = append(errors, error.AsMap())
+	}
+
+	return map[string]interface{} {
+		Description: problem.title,
```
Comment (drsnyder) [inline problem/problem.go]: Should this be `Title`?

Comment (nmonterroso) [inline problem/problem.go]: oops af205f9

File: controllers/register_controller_e2e_test.go (hunk 4 above)
Comment (nmonterroso) [inline controllers/register_controller_e2e_test.go]: Actually, these ints make the string too big, so gonna stick with time for now.

**Comment by drsnyder** [comment]
:sheep: it

**Comment by ArturKlajnerok** [comment]
Reviewed it after merge and LGTM :smile_cat:
